elastic/apm-server
# Refactor metadata enrichment and fix bug in v3 Intake

## Motivation/summary
Decode request metadata into struct instead of a map and pass metadata struct
into decoder function. The refactoring automatically fixes a bug where client.ip
and user_agent was not set for v3 RUM events.

With these changes the `host.ip`  is now also added to the `metricset` events.

## Checklist

~- [ ] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).~

I have considered changes for:
~- [ ] documentation~
~- [ ] logging (add log lines, choose appropriate log selector, etc.)~
~- [ ] metrics and monitoring (create issue for Kibana team to add metrics to visualizations, e.g. [Kibana#44001](https://github.com/elastic/kibana/issues/44001))~
~- [ ] telemetry~
~- [ ] Elasticsearch Service (https://cloud.elastic.co)~
~- [ ] Elastic Cloud Enterprise (https://www.elastic.co/products/ece)~
~- [ ] Elastic Cloud on Kubernetes (https://www.elastic.co/elastic-cloud-kubernetes)~

## How to test these changes
Run APM Server with RUM enabled, send RUM request and check that documents in ES are enriched with `client.ip`  and `user_agent`. 

## Related issues
implements #3885
fixes #3942
